backend/danswer/llm/answering/stream_processing/citation_processing.py

```diff
@@ -26,91 +26,161 @@ def extract_citations_from_stream(
     doc_id_to_rank_map: dict[str, int],
     stop_stream: str | None = STOP_STREAM_PAT,
 ) -> Iterator[DanswerAnswerPiece | CitationInfo]:
+    """
+    Key aspects:
+
+    1. Stream Processing:
+    - Processes tokens one by one, allowing for real-time handling of large texts.
+
+    2. Citation Detection:
+    - Uses regex to find citations in the format [number].
+    - Example: [1], [2], etc.
+
+    3. Citation Mapping:
+    - Maps detected citation numbers to actual document ranks using doc_id_to_rank_map.
+    - Example: [1] might become [3] if doc_id_to_rank_map maps it to 3.
+
+    4. Citation Formatting:
+    - Replaces citations with properly formatted versions.
+    - Adds links if available: [[1]](https://example.com)
+    - Handles cases where links are not available: [[1]]()
+
+    5. Duplicate Handling:
+    - Skips consecutive citations of the same document to avoid redundancy.
+
+    6. Output Generation:
+    - Yields DanswerAnswerPiece objects for regular text.
+    - Yields CitationInfo objects for each unique citation encountered.
+
+    7. Context Awareness:
+    - Uses context_docs to access document information for citations.
+
+    This function effectively processes a stream of text, identifies and reformats citations,
+    and provides both the processed text and citation information as output.
+    """
+
     llm_out = ""
     max_citation_num = len(context_docs)
     curr_segment = ""
-    prepend_bracket = False
     cited_inds = set()
     hold = ""
+
+    raw_out = ""
+    current_citations = []
+    past_cite_count = 0
     for raw_token in tokens:
+        raw_out += raw_token
         if stop_stream:
             next_hold = hold + raw_token
-
             if stop_stream in next_hold:
                 break
-
             if next_hold == stop_stream[: len(next_hold)]:
                 hold = next_hold
                 continue
-
             token = next_hold
             hold = ""
         else:
             token = raw_token
 
-        # Special case of [1][ where ][ is a single token
-        # This is where the model attempts to do consecutive citations like [1][2]
-        if prepend_bracket:
-            curr_segment += "[" + curr_segment
-            prepend_bracket = False
-
         curr_segment += token
         llm_out += token
 
-        possible_citation_pattern = r"(\[\d*$)"  # [1, [, etc
```

Comment:
if we're no longer looking for these possible citation patterns, we could now output partial citations? E.g.
if the stream is like `["[1", "]."]`, then the `[1` will flash in the UI before being converted into a citation. This is probably even more important once we switch over to the new chat UI with the special display for citations